### PR TITLE
Added support for changing value type in changedKeys (fixes #982).

### LINKED ISSUE #982: [QUESTION]: Can't convert string to object
Hi, I seem to be experiencing a weird issue.

I am using a GraphQL schema with my form and I have a field that is a String type.
In my model, I load the string with a value.
I have a Select that is a custom component that allows me to select a different value and the selected value returned is an object.

When I change the selection the field in the model is updated correctly from a string to an object, and in the transform I can change it when I submit my form.

The issue I am having is when I use the formRef that I passed to AutoForm, and call onChange to change my field value to an object, it gives me this error: 

> Cannot use 'in' operator to search for 'name' in some string

It might be related to: https://github.com/vazco/uniforms/issues/960

You can see an example here:

https://codesandbox.io/s/cant-convert-string-to-object-with-onchange-o1o7q

I am using latest uniforms version.

I am triggering the change with `useEffect` in the code on line `116`.

In this example, I did not put a Select as the field, I simply wanted to show the error message.

Hope you can help.
Let me know if I am not being clear in my description.

Thanks!

### PR DESCRIPTION
**Pull request by simplecommerce**
added typeOf valueA and valueB for changedKeys

PR for https://github.com/vazco/uniforms/issues/982#issuecomment-898503298

### CODE REVIEW

File: packages/uniforms/src/changedKeys.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { joinName } from './joinName';
 
 // eslint-disable-next-line complexity
 export function changedKeys<T>(root: string, valueA?: T, valueB?: T) {
-  if (!valueA || valueA !== Object(valueA) || valueA instanceof Date) {
+  if (!valueA || typeof valueA !== typeof valueB || valueA !== Object(valueA) || valueA instanceof Date) {
```
Comment (radekmie) [inline packages/uniforms/src/changedKeys.ts:7]: As it's going to be the most uncommon case, I'd move it to the end of this condition:
```suggestion
  if (!valueA || valueA !== Object(valueA) || valueA instanceof Date || typeof valueA !== typeof valueB) {
```

**Review by radekmie — CHANGES_REQUESTED** [review]
Thanks for the PR! We'd need to add tests though. In #982 you've said there was some problem on your end. Have you tried the steps described in [CONTRIBUTING.md](https://github.com/vazco/uniforms/blob/master/.github/CONTRIBUTING.md#making-changes)? Basically `npm ci` and `npm test` in the root should be enough. If there'll be any problems, let us know in the comments here.

**Comment by simplecommerce** [comment]
> Thanks for the PR! We'd need to add tests though. In #982 you've said there was some problem on your end. Have you tried the steps described in [CONTRIBUTING.md](https://github.com/vazco/uniforms/blob/master/.github/CONTRIBUTING.md#making-changes)? Basically `npm ci` and `npm test` in the root should be enough. If there'll be any problems, let us know in the comments here.

Hi thanks for the help, yes I tried that.
I added some tests in my PR but when I do `npm test` after `npm ci` i get this error now.

```
sh: 1: eslint: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! uniforms-repository@ lint:code: `eslint --cache --cache-location node_modules/.cache/eslint --cache-strategy content --ext js,ts,tsx .`
npm ERR! spawn ENOENT
npm ERR!
npm ERR! Failed at the uniforms-repository@ lint:code script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-08-16T09_38_03_428Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! uniforms-repository@ lint: `npm run lint:code && npm run lint:text && npm run lint:types`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the uniforms-repository@ lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-08-16T09_38_03_447Z-debug.log
npm ERR! Test failed.  See above for more details.
```

**Comment by radekmie** [comment]
Can you make sure that you're using npm v7? It looks like you're not.

**Comment by simplecommerce** [comment]
> Can you make sure that you're using npm v7? It looks like you're not.

Hi, you were right I needed to update npm to v7! 
It seems to work now, I added some tests and when I run them I get this error.
I am not familiar with TypeScript so I don't know how to fix this.

```
packages/uniforms/__tests__/changedKeys.ts:100:37 - error TS2345: Argument of type '{ test: boolean; }' is not assignable to parameter of type 'string'.

100     expect(changedKeys('a', 'test', { test: true })).toEqual(['a']);
                                        ~~~~~~~~~~~~~~

packages/uniforms/__tests__/changedKeys.ts:101:37 - error TS2345: Argument of type '1' is not assignable to parameter of type 'string | undefined'.

101     expect(changedKeys('a', 'test', 1)).toEqual(['a']);
                                        ~

packages/uniforms/__tests__/changedKeys.ts:102:37 - error TS2345: Argument of type 'true' is not assignable to parameter of type 'string | undefined'.

102     expect(changedKeys('a', 'test', true)).toEqual(['a']);
                                        ~~~~

packages/uniforms/__tests__/changedKeys.ts:103:51 - error TS2554: Expected 1-3 arguments, but got 4.

103     expect(changedKeys('a', 'test', new Date(10), new Date(20))).toEqual(['a']);
                                                      ~~~~~~~~~~~~

packages/uniforms/__tests__/changedKeys.ts:104:37 - error TS2345: Argument of type 'string[]' is not assignable to parameter of type 'string'.

104     expect(changedKeys('a', 'test', ["test"])).toEqual(['a']);
                                        ~~~~~~~~

packages/uniforms/__tests__/changedKeys.ts:105:29 - error TS2345: Argument of type 'string[]' is not assignable to parameter of type 'string'.

105     expect(changedKeys('a', ['test'], "test")).toEqual(['a']);
                                ~~~~~~~~

packages/uniforms/__tests__/changedKeys.ts:107:29 - error TS2345: Argument of type 'string[]' is not assignable to parameter of type 'number'.

107     expect(changedKeys('a', ['test'], 1)).toEqual(['a']);
                                ~~~~~~~~

packages/uniforms/__tests__/changedKeys.ts:108:29 - error TS2345: Argument of type 'string[]' is not assignable to parameter of type 'boolean | undefined'.
  Type 'string[]' is not assignable to type 'true'.

108     expect(changedKeys('a', ['test'], true)).toEqual(['a']);
                                ~~~~~~~~

packages/uniforms/__tests__/changedKeys.ts:109:53 - error TS2554: Expected 1-3 arguments, but got 4.

109     expect(changedKeys('a', ['test'], new Date(10), new Date(20))).toEqual(['a']);
                                                        ~~~~~~~~~~~~
```

**Comment by radekmie** [comment]
@simplecommerce The problem is that now `changedKeys` requires that both `valueA` and `valueB` are of the same type. I'll update your branch with a fix and corresponding tests later this week.

**Review by radekmie — APPROVED** [review]

**Comment by radekmie** [comment]
@wadamek65, @Monteth, @kestarumper Could you review this PR?

**Review by kestarumper — APPROVED** [review]

**Pull request merged**